### PR TITLE
Volume, play/pause, and fix sink dropping

### PR DESCRIPTION
**Pull request by Xaeroxe23**
Fixed the set_volume function of a Sink and adds a get_volume function.

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -35,7 +35,15 @@ pub struct Engine {
 struct EndPointVoices {
     format: Format,
     next_id: AtomicUsize,
-    pending_sounds: Mutex<Vec<(usize, QueueIterator)>>,
+    pending_sounds: Mutex<Vec<HandleHandle>>,
+}
+
+// Handle to a handle I guess.
```
Comment (tomaka) [inline src/engine.rs]: Same two remarks as #84

Comment (Xaeroxe23) [inline src/engine.rs]: Ok, there is no such thing as an AtomicF32 though so I assume the volume level should still be a Mutex<f32>

**Comment by Xaeroxe23** [comment]
The other two have been merged into this request and the requested changes have been made.

**Comment by Xaeroxe23** [comment]
@tomaka Ready for review.

**Comment by tomaka** [comment]
Please load the volume in a temporary variable outside of the iterator (for example between lines 126 and 127) so that we only lock the mutex once from time to time. That should be okay-ish I think.

File: src/lib.rs [hunk 2]
```diff
@@ -103,6 +103,34 @@ impl Sink {
         self.handle.append(source);
     }
 
+    /// Gets the volume of the sound.
+    ///
+    /// The value `1.0` is the "normal" volume (unfiltered input). Any value other than 1.0 will
+    /// multiply each sample by this value.
+    #[inline]
+    pub fn get_volume(&self) -> f32 {
```
Comment (tomaka) [inline src/lib.rs]: Please rename this to just `volume()`.

File: src/engine.rs [hunk 3]
```diff
@@ -206,18 +250,42 @@ impl Handle {
         self.next_sounds.lock().unwrap().push((source, Some(tx)));
     }
 
+    /// Gets the volume of the sound played by this sink.
+    #[inline]
+    pub fn get_volume(&self) -> f32 {
```
Comment (tomaka) [inline src/engine.rs]: Same here.

**Comment by tomaka** [comment]
Sorry, I didn't review your volume PR carefully. This time it should be good after these changes.

**Comment by Xaeroxe23** [comment]
Ok, so I agree it is not desirable for the sampling thread to block, locking can lead to blocking.  However I do want the volume to be fetched for every sample as that makes the set_volume function feel the most responsive.  It's not possible to make the f32 atomic.  So the approach I'm currently thinking of would use a [channel](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.htmll)<f32>, we'd store the Sender in the Handle and the Receiver in the HandleHandle, then whenever the Handle wishes to change the volume level it will send an f32 through the channel, HandleHandle would store an f32 called volume and in the sample iterator (lines 128-133) we would call try_recv to fetch any new volume level sent by Handle, and store it in HandleHandle.  This will **not** be contained in a while loop, it will only attempt to receive one volume level per sample, which prevents the volume thread from being DDOSed by a large influx of volume change requests.

I'll create and push an implementation of this on my lunch break, let me know if you find any fault in it.

**Comment by tomaka** [comment]
I don't think it's better to use a channel.

The iterator that builds the samples is a code that is very easy to optimize for the compiler. The fact that we're using the `Relaxed` ordering means that in theory it can fetch the value of "paused" only from time to time and build the samples through SIMD for example.

If you're introducing a channel read or a mutex in there, you're not only paying the price of the channel or mutex read, but you're also totally killing any potential for a compiler optimisation on this code.

In fact I have already doubts about the fact that `paused` doesn't do kill all optimisations already, but I thought it would be "good enough" while waiting for a more robust rewrite of the engine's code.

**Comment by Xaeroxe23** [comment]
I have a difficult time imagining how the engine could have the features in this request in a more optimized manner, unless we were to make changes to cpal.  That being said, currently on line 126 of engine.rs I am dropping the sound if it was declared dead, and that seems to be responsive enough in my own testing, so on second thought I'm not opposed to loading temporary volume variables here.

**Comment by tomaka** [comment]
> I have a difficult time imagining how the engine could have the features in this request in a more optimized manner, unless we were to make changes to cpal.

You could put the `Arc<Mutex<f32>>` of the volume in the source of samples itself, alongside with a local copy of the volume. The source of samples would then load the value from the mutex to its local copy from time to time.

**Comment by Xaeroxe23** [comment]
Ok, that makes sense.  I'd like to see if I can accomplish that, should "from time to time" just be approximately every 5ms?

**Comment by Xaeroxe23** [comment]
Some context: I'm using rodio in a game engine project and I want to give users of my engine more control over the playback of their sounds, those features that I need to do that are in this PR.  I don't have anywhere near the same level of experience manipulating audio devices as you do but I gave it my best attempt.  I'm trying to avoid having an audio engine that is just "good enough for now" I'd like it to be at this ideal performance level you describe.  So I'm willing to tackle the idea of finding a good place and way to do more infrequent fetching of the variables into a local copy.

**Comment by Xaeroxe23** [comment]
Alternatively if you want to do a rewrite of the engine I could just add my temporary variable after line 126 as you originally suggested and just be done with it.

**Comment by tomaka** [comment]
I'd like to rewrite some parts of the engine but I don't know if it's going to happen in the next month or in the next year.

Basically the ideal solution would be to add a filter (similar to the ones in `src/source`) whose constructor takes a source and an `Arc<Mutex<f32>>`.
The filter would calculate approximately the number of samples that are equal to 5 ms, and re-read from the mutex once that number of samples has been processed.
The actual duration of 5ms could perhaps be customized with another constructor.

Then the engine could wrap all sound sources passed to it around that filter.

**Comment by Xaeroxe23** [comment]
Ok, do you also feel we should be doing something similar with the AtomicBools or is the current implementation sufficient for those?

**Comment by tomaka** [comment]
@Xaeroxe23 Yeah I think that would be better.
Once you have written one of the filters, adding the others should be very easy after some copy-pasting.

**Comment by Xaeroxe23** [comment]
It just occurred to me dead doesn't modify the sample output though so that's only really applicable to paused.

**Comment by Xaeroxe23** [comment]
Resolved in another PR, closing.

**Pull request closed without merging**